Automated Smoketest: :x:
OS: centos 7.2.1511 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: 48079747effc6c985c51583684f898711f0a4450
Number of Error(s): 7
 /usr/bin/install -c -m 644 include/antlr3.h include/antlr3baserecognizer.h include/antlr3basetree.h include/antlr3basetreeadaptor.h include/antlr3bitset.h include/antlr3collections.h include/antlr3commontoken.h include/antlr3commontree.h include/antlr3commontreeadaptor.h include/antlr3commontreenodestream.h include/antlr3convertutf.h include/antlr3cyclicdfa.h include/antlr3debugeventlistener.h include/antlr3defs.h include/antlr3encodings.h include/antlr3errors.h include/antlr3exception.h include/antlr3filestream.h include/antlr3input.h include/antlr3interfaces.h include/antlr3intstream.h include/antlr3lexer.h include/antlr3memory.h include/antlr3parser.h include/antlr3parsetree.h include/antlr3recognizersharedstate.h include/antlr3rewritestreams.h include/antlr3string.h include/antlr3tokenstream.h include/antlr3treeparser.h antlr3config.h /mnt/disk1/home/vamosax/smoketest/PR-10932/build/esp/services/ws_sql/antlr3c/include
[  0%] Building CXX object aws-cpp-sdk-core/CMakeFiles/aws-cpp-sdk-core.dir/source/client/CoreErrors.cpp.o
[100%] Building CXX object aws-cpp-sdk-sqs/CMakeFiles/aws-cpp-sdk-sqs.dir/source/SQSErrors.cpp.o
collect2: error: ld returned 1 exit status
make[2]: *** [Release/bin/eclcc] Error 1
make[1]: *** [ecl/eclcc/CMakeFiles/eclcc.dir/all] Error 2
make: *** [all] Error 2
Get ECLWatch build result

[Build error:](https://gist.github.com/4617c360f3c898bdaec3c4d8ac12aa10)
Build log [Build log](https://gist.github.com/HPCCSmoketest/4617c360f3c898bdaec3c4d8ac12aa10#file-result-18-03-13-13-00-28-log)

--- a/common/fileview2/fvtransform.cpp
+++ b/common/fileview2/fvtransform.cpp
@@ -297,8 +297,9 @@ void ViewTransformerRegistry::addPlugins(const char * name)
     Owned<IErrorReceiver> errorReporter = createThrowingErrorReceiver();
     dataServer.setown(createNewSourceFileEclRepository(errorReporter, name, ESFallowplugins, 0));
 
+    NullStatisticTarget nullStats;
     HqlScopeArray scopes;
-    HqlParseContext parseCtx(dataServer, NULL, NULL);
+    HqlParseContext parseCtx(dataServer, NULL, NULL, nullStats);
     HqlLookupContext ctx(parseCtx, errorReporter);
     getRootScopes(scopes, dataServer, ctx);
 

--- a/ecl/eclcc/eclcc.cpp
+++ b/ecl/eclcc/eclcc.cpp
@@ -1201,11 +1201,14 @@ void EclCC::processSingleQuery(EclCompileInstance & instance,
 
     {
         //Minimize the scope of the parse context to reduce lifetime of cached items.
-        HqlParseContext parseCtx(instance.dataServer, &instance, instance.archive);
+        WuStatisticTarget statsTarget(instance.wu, "eclcc");
+        HqlParseContext parseCtx(instance.dataServer, &instance, instance.archive, statsTarget);
         parseCtx.cache = cache;
         parseCtx.optionHash = optionHash;
         if (optFastSyntax)
             parseCtx.setFastSyntax();
+        parseCtx.timeParser = instance.wu->getDebugValueBool("timeParser", false);
+
         unsigned maxErrorsDebugOption = instance.wu->getDebugValueInt("maxErrors", 0);
         if (maxErrorsDebugOption != 0)
             parseCtx.maxErrors = maxErrorsDebugOption;

--- a/ecl/eclcc/eclcc.hpp
+++ b/ecl/eclcc/eclcc.hpp
@@ -140,6 +140,7 @@ const char * const helpText[] = {
     "?!  -fshowRecordCountInGraph  Show estimates of record counts in the graph",
     "?!  -fspanMultipleCpp       Generate a work unit in multiple c++ files",
     "?!  -fsubgraphToRegenerate=n Regenerate the ECL for a particular subgraph",
+    "?!  -ftimeParser            Add timings for parsing each ECL attribute",
     "?!  -ftimeTransforms        Add timings for internal transforms to the workunit",
     "",
 };

--- a/ecl/hql/hqlexpr.cpp
+++ b/ecl/hql/hqlexpr.cpp
@@ -1345,6 +1345,13 @@ extern HQL_API IPropertyTree * queryArchiveEntry(IPropertyTree * archive, const 
 
 //---------------------------------------------------------------------------------------------------------------------
 
+HqlLookupContext::~HqlLookupContext()
+{
+    //If timings were not accumulated for this lookup context, accumulate child timings in the parent
+    if (container)
+        container->childCycles += childCycles;
+}
+
 void HqlLookupContext::noteBeginAttribute(IHqlScope * scope, IFileContents * contents, IIdAtom * name)
 {
     parseCtx.noteBeginAttribute(scope, contents, name);
@@ -1362,6 +1369,27 @@ void HqlLookupContext::noteBeginModule(IHqlScope * scope, IFileContents * conten
     parseCtx.noteBeginModule(scope, contents);
 }
 
+
+void HqlLookupContext::reportTiming(const char * name)
+{
+    if (startCycles)
+    {
+        cycle_t endCycles = get_cycles_now();
+        cycle_t elapsedCycles = endCycles - startCycles;
+        if (container)
+            container->childCycles += elapsedCycles;
+
+        __uint64 totalTime = cycle_to_nanosec(elapsedCycles);
+        __uint64 localTime = cycle_to_nanosec(elapsedCycles - childCycles);
+
+        StringBuffer scope;
+        scope.append("compiler::parse::").append(name);
+        parseCtx.statsTarget.addStatistic(SSTcompilestage, scope, StTimeTotalExecute, nullptr, totalTime, 1, 0, StatsMergeSum);
+        parseCtx.statsTarget.addStatistic(SSTcompilestage, scope, StTimeLocalExecute, nullptr, localTime, 1, 0, StatsMergeSum);
+
+        childCycles = 0;
+    }
+}
 
 //---------------------------------------------------------------------------------------------------------------------
 
@@ -16912,7 +16940,8 @@ static void gatherAttributeDependencies(HqlLookupContext & ctx, const char * ite
 
 extern HQL_API IPropertyTree * gatherAttributeDependencies(IEclRepository * dataServer, const char * items)
 {
-    HqlParseContext parseCtx(dataServer, NULL, NULL);
+    NullStatisticTarget nullStats;
+    HqlParseContext parseCtx(dataServer, NULL, NULL, nullStats);
     parseCtx.nestedDependTree.setown(createPTree("Dependencies"));
 
     Owned<IErrorReceiver> errorHandler = createNullErrorReceiver();

--- a/ecl/hql/hqlexpr.hpp
+++ b/ecl/hql/hqlexpr.hpp
@@ -874,8 +874,8 @@ public:
         StringAttr cacheLocation;
     };
 
-    HqlParseContext(IEclRepository * _eclRepository, ICodegenContextCallback *_codegenCtx, IPropertyTree * _archive)
-    : archive(_archive), eclRepository(_eclRepository), codegenCtx(_codegenCtx)
+    HqlParseContext(IEclRepository * _eclRepository, ICodegenContextCallback *_codegenCtx, IPropertyTree * _archive, IStatisticTarget & _statsTarget)
+    : archive(_archive), eclRepository(_eclRepository), codegenCtx(_codegenCtx), statsTarget(_statsTarget)
     {
         expandCallsWhenBound = true;
         ignoreUnknownImport = false;
@@ -924,6 +924,7 @@ public:
     IErrorArray orphanedWarnings;
     HqlExprArray defaultFunctionCache;
     CIArrayOf<ForwardScopeItem> forwardLinks;
+    IStatisticTarget & statsTarget;
     unsigned maxErrors = DEFAULT_MAX_ERRORS;
     bool unsuppressImmediateSyntaxErrors = false;
     bool expandCallsWhenBound;
@@ -931,6 +932,7 @@ public:
     bool ignoreSignatures;
     bool aborting;
     bool checkDirty = false;
+    bool timeParser = false;
     Linked<ICodegenContextCallback> codegenCtx;
     CIArrayOf<FileParseMeta> metaStack;
     IEclCachedDefinitionCollection * cache = nullptr;
@@ -953,23 +955,31 @@ private:
 class HqlDummyParseContext : public HqlParseContext
 {
 public:
-    HqlDummyParseContext() : HqlParseContext(NULL, NULL, NULL) {}
+    HqlDummyParseContext() : HqlParseContext(NULL, NULL, NULL, nullStats) {}
+
+    NullStatisticTarget nullStats;
 };
 
 
-class HqlLookupContext
+class HQL_API HqlLookupContext
 {
 public:
-    explicit HqlLookupContext(const HqlLookupContext & other) : parseCtx(other.parseCtx)
+    explicit HqlLookupContext(HqlLookupContext & other) : parseCtx(other.parseCtx)
     {
         errs.set(other.errs); 
-        functionCache = other.functionCache; 
+        functionCache = other.functionCache;
+        container = &other;
+        if (parseCtx.timeParser)
+            startCycles = get_cycles_now();
     }
     HqlLookupContext(HqlParseContext & _parseCtx, IErrorReceiver * _errs)
     : parseCtx(_parseCtx), errs(_errs)
     {
         functionCache = &parseCtx.defaultFunctionCache;
+        if (parseCtx.timeParser)
+            startCycles = get_cycles_now();
     }
+    ~HqlLookupContext();
 
     void noteBeginAttribute(IHqlScope * scope, IFileContents * contents, IIdAtom * name);
     void noteBeginModule(IHqlScope * scope, IFileContents * contents);
@@ -992,6 +1002,8 @@ public:
     inline void setAborting() { parseCtx.setAborting(); }
     inline bool checkDirty() const { return parseCtx.checkDirty; }
 
+    void reportTiming(const char * name);
+
 protected:
 
     inline IPropertyTree * queryArchive() const { return parseCtx.archive; }
@@ -1003,6 +1015,9 @@ private:
 public:
     Linked<IErrorReceiver> errs;
     HqlExprArray * functionCache;
+    cycle_t startCycles = 0;
+    cycle_t childCycles = 0;
+    HqlLookupContext * container = nullptr;
 };
 
 class HqlDummyLookupContext : public HqlLookupContext

--- a/ecl/hql/hqlgram2.cpp
+++ b/ecl/hql/hqlgram2.cpp
@@ -12207,6 +12207,16 @@ void parseAttribute(IHqlScope * scope, IFileContents * contents, HqlLookupContex
     bool isMacro = parsed && parsed->isMacro();
 
     attrCtx.noteEndAttribute(true, canCache, isMacro, simplified);
+
+    if (attrCtx.queryParseContext().timeParser)
+    {
+        StringBuffer fullname;
+        fullname.append(scope->queryFullName());
+        if (fullname)
+            fullname.append(".");
+        fullname.append(str(name));
+        attrCtx.reportTiming(fullname);
+    }
 }
 
 int testHqlInternals()

--- a/ecl/hql/hqlplugininfo.cpp
+++ b/ecl/hql/hqlplugininfo.cpp
@@ -86,7 +86,8 @@ static void expandPluginPropertyTree(IPropertyTree * target, HqlLookupContext & 
 
 IPropertyTree * createPluginPropertyTree(IEclRepository * plugins, bool includeModuleText)
 {
-    HqlParseContext parseCtx(plugins, NULL, NULL);
+    NullStatisticTarget nullStats;
+    HqlParseContext parseCtx(plugins, NULL, NULL, nullStats);
     HqlLookupContext ctx(parseCtx, NULL);
 
     Owned<IPropertyTree> map = createPTree("Plugins", ipt_caseInsensitive);
@@ -107,7 +108,8 @@ IPropertyTree * getPlugin(IPropertyTree * p, IEclRepository * plugins, const cha
 
     if(load && !plugin->getPropInt("@loaded",0))
     {
-        HqlParseContext parseCtx(plugins, NULL, NULL);
+        NullStatisticTarget nullStats;
+        HqlParseContext parseCtx(plugins, NULL, NULL, nullStats);
         HqlLookupContext GHMOREctx(parseCtx, NULL);
         Owned<IHqlScope> resolved = getResolveDottedScope(modname, LSFpublic, GHMOREctx);
         if (resolved)

--- a/system/jlib/jstats.h
+++ b/system/jlib/jstats.h
@@ -729,4 +729,13 @@ interface IStatisticTarget
     virtual void addStatistic(StatisticScopeType scopeType, const char * scope, StatisticKind kind, char * description, unsigned __int64 value, unsigned __int64 count, unsigned __int64 maxValue, StatsMergeAction mergeAction) = 0;
 };
 
+class jlib_decl NullStatisticTarget : implements IStatisticTarget
+{
+public:
+    virtual void addStatistic(StatisticScopeType scopeType, const char * scope, StatisticKind kind, char * description, unsigned __int64 value, unsigned __int64 count, unsigned __int64 maxValue, StatsMergeAction mergeAction)
+    {
+    }
+};
+
+
 #endif

